Add methods for dividing triangular by diagonal matrices.

tpapp opened the pull request:
Fixes JuliaLang/LinearAlgebra.jl#541.

tpapp committed:
Add methods for dividing triangular by diagonal matrices.

Fixes #27989.

fredrikekre commented on stdlib/LinearAlgebra/test/diagonal.jl:
```diff
@@ -455,4 +455,21 @@ end
 end
 
+@testset "Triangular division by Diagonal #27989" begin
+    for _ in 1:100
```
This loop seems unnecessary?

tpapp replied:
You mean that testing with multiple random values is unnecessary, and I should go for a single one?

The major cost is compilation, once compiled, additional random tests are relatively cheap.

fredrikekre replied:
Right, but we don't do this anywhere else. For testing correctness of the code path it would be more useful with a loop over e.g. `(Float64, ComplexF64)`.

tpapp replied:
Thanks for the suggestion, this actually caught a bug.

fredrikekre commented on stdlib/LinearAlgebra/test/diagonal.jl:
```diff
@@ -455,4 +455,21 @@ end
+        L = LowerTriangular(randn(K, K))
+        D = Diagonal(randn(K))
+        @test U / D isa UpperTriangular
+        @test U / D == UpperTriangular(Matrix(U) / Matrix(D))
```
Perhaps combine these two (and similar below) as
```julia
@test (U / D)::UpperTriangular == UpperTriangular(Matrix(U) / Matrix(D))
```
for the same effect?

tpapp committed:
Simplified tests.

suggested by @fredrikekre

tpapp committed:
Test for element types instead, fix bug.

For broadcasting, tranpose should be used instead of adjoint.

andreasnoack commented on stdlib/LinearAlgebra/src/diagonal.jl:
```diff
@@ -356,6 +356,11 @@ rdiv!(A::AbstractMatrix{T}, transD::Transpose{<:Any,<:Diagonal{T}}) where {T} =
 (\)(A::Union{QR,QRCompactWY,QRPivoted}, B::Diagonal) =
     invoke(\, Tuple{Union{QR,QRCompactWY,QRPivoted}, AbstractVecOrMat}, A, B)
 
+(/)(U::UpperTriangular, D::Diagonal) = UpperTriangular(parent(U) ./ transpose(D.diag))
```
It would be great if you could define the in-place versions `ldiv!` and `rdiv!` instead and then adjust the methods in line 427-429 to use these.

andreasnoack replied:
Probably just something like
```julia
function rdiv!(U::UpperTriangular, D::Diagonal)
broadcast!(/, parent(U), parent(U), diag(D))
return U
```
although using broadcast here is actually cheating a bit since broadcasting and linear algebra operations have slightly different semantics in corner cases. However, that is an existing issue so we can clean that up in another PR.

andreasnoack replied:
Hm. The `transpose` is actually also not correct here since it is recursive so you probably want to reshape the vector instead.

andreasnoack replied:
We'd need the `rdiv!` and `ldiv!` methods anyway and it would be confusing if the non-mutating versions didn't go through the mutating versions. Also, the promotion can be done in the more generic version so it's not that much code. Good idea to use broadcast of the type for that.

tpapp committed:
Matrix diagonal division through ldiv!/rdiv!.

Removed previous direct implementations for upper/lower triangular,
added ldiv!/rdiv! methods.

Use reshape instead of transpose.

tpapp commented:
I did not run the full test suite before committing, and now the (freebsd) tests show an ambiguity which I expect will crop up in the other CI runs, too.

Suggestions on how to handle this would be appreciated.

andreasnoack commented:
You'd need to add definitions for
```julia
(/)(A::Transpose, D::Diagonal)
(/)(A::Adjoint, D::Diagonal)
(/)(x::RowVector, D::Diagonal)
```
where the latter should be defined as a deprecated method. The two former might be joined to `AdjOrTrans` but I'm not sure.

tpapp commented:
Thanks, I am aware that defining the intersection is the standard solution for ambiguities, but I was wondering if there is a better method for streamlining the architecture.

I am wondering about rolling back to 985891d95c98783b7ab36bc2863ae6ee0be88b88, replacing the `transpose` with `reshape`, and leaving the `rdiv!` and `ldiv!` for future work.

I am also skeptical about whether we need an `ldiv!` and `rdiv!` method  at all, since eg `ldiv!` says that

> The argument `A` should *not* be a matrix

and `UpperTriangular` and `LowerTriangular` are matrices.

andreasnoack commented:
I'm also not too happy about all the need for disambiguating methods. Maybe just restrict `(/)(A::AbstractMatrix, D::Diagonal)` to `(/)(A::StridedMatrix, D::Diagonal)`. That will make the `AbstractMatrix` case go through the fallback that calls `\` but I think that is fine.

Regarding the formulation in the docstring for `ldiv!` then it is not precise but it doesn't change the fact that that `\` and `/` should go through `ldiv!` and `rdiv!`. The reason for the formulation is that it should be possible to solve directly with the first argument in `ldiv!`. That typically means that a matrix should be in factored form but that is, of course, not necessary for a `Diagonal`.

andreasnoack commented on stdlib/LinearAlgebra/src/diagonal.jl:
```diff
@@ -325,6 +325,11 @@ ldiv!(adjD::Adjoint{<:Any,<:Diagonal{T}}, B::AbstractVecOrMat{T}) where {T} =
     (D = transD.parent; ldiv!(D, B))
 
+function ldiv!(D::Diagonal, A::Union{LowerTriangular,UpperTriangular})
+    broadcast!(/, parent(A), parent(A), diag(D))
```
I forgot the reshape here. Needs for be `reshape(D.diag, 1, length(D.diag))` instead. (I recently learned that `diag` now copies.)

tpapp replied:
Happy to add this, but can you please explain why `reshape` is needed here? In any case it should be
```julia
reshape(D.diag, length(D.diag), 1) 
```
for the `ldiv!`, but wouldn't using just the vector `D.diag` be equivalent?

andreasnoack replied:
I saw you were using `/` and thought this was `rdiv!`. This should then be
```julia
broadcast!(\, parent(A), D.diag, parent(A))
```
`d\a` and `a/d` is not always the same. E.g. if the elements are matrices.

tpapp committed:
Fix ldiv! to \, use permutedims, restrict signature.

tpapp commented:
Thanks for all the help!
